reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java

```diff
@@ -338,23 +403,29 @@ public boolean wasRequested() {
 
 	@Override
 	public ColdTestPublisher<T> assertRequestOverflow() {
-		//the cold publisher cannot really overflow requests, as it immediately throws
-		throw new AssertionError("Expected some request overflow");
+		if (!hasOverflown) {
+			throw new AssertionError("Expected some request overflow");
+		}
+		return this;
 	}
 
 	@Override
 	public ColdTestPublisher<T> assertNoRequestOverflow() {
-		//the cold publisher cannot really overflow requests, as it immediately throws
+		if (hasOverflown) {
+			throw new AssertionError("Unexpected request overflow");
+		}
 		return this;
 	}
 
 	@Override
 	public ColdTestPublisher<T> next(@Nullable T t) {
-		Objects.requireNonNull(t, "emitted values must be non-null");
+		if (!violations.contains(ALLOW_NULL)) {
+			Objects.requireNonNull(t, "emitted values must be non-null");
+		}
 
 		values.add(t);
 		for (ColdTestPublisherSubscription<T> s : subscribers) {
-			s.onNext(t);
+			s.drain(s.requested);
```

Comment:
wouldn't it be possible for the `drain` to have no parameter, computing the `n` from its latest `requested` state?

Reply:
I've always been puzzled by that myself everywhere we see this construct, but I guess for the case where it is called from `request(n)` we can't, can we?
```java
		public void request(long n) {
			if (Operators.validate(n)) {
				if (Operators.addCap(REQUESTED, this, n) == 0) {
					parent.wasRequested = true;
					drain(n);
				}
			}
		}
```

Reply:
Actually, we can, since it is only called if total request *now* equals `n`. Hmmm 🤔